src/TBuffer.cpp

```diff
@@ -4413,11 +4422,128 @@ bool TBuffer::processGBSequence(const std::string& bufferData, const bool isFrom
     return true;
 }
 
+bool TBuffer::processBig5Sequence(const std::string& bufferData, const bool isFromServer, const size_t len, size_t& pos, bool& isNonBmpCharacter)
+{
+#if defined(DEBUG_BIG5_PROCESSING)
+    std::string dataIdentity;
+#endif
+
+    // The range deductions for two-byte sequences are take from:
+    // https://en.wikipedia.org/wiki/GBK#Encoding
+    size_t big5SequenceLength = 1;
+    bool isValid = true;
+    bool isToUseReplacementMark = false;
+    // Only set this if we are adding more than one code-point to
+    // mCurrentLineCharacters:
+    isNonBmpCharacter = false;
+    if (static_cast<quint8>(bufferData.at(pos)) < 0x80) {
+        // Is ASCII - single byte character, straight forward for a "first" byte case
+        mMudLine.append(bufferData.at(pos));
+        // As there is already a unit increment at the bottom of caller's loop
+        // there is no need to tweak pos in THIS case
+
+        return true;
+    } else if (static_cast<quint8>(bufferData.at(pos)) == 0x80 || static_cast<quint8>(bufferData.at(pos)) > 0xFE) {
+        // Invalid as first byte
+        isValid = false;
+        isToUseReplacementMark = true;
+#if defined(DEBUG_BIG5_PROCESSING)
+        qDebug().nospace() << "TBuffer::processBig5Sequence(...) 1-byte sequence as Big5 rejected!";
+#endif
+
+        // Proceed to handle 1 byte (as GB2312/GBK/GB18030 data) outside of checks...
+
+    } else {
+        // We have two bytes
+        big5SequenceLength = 2;
+        if ((pos + big5SequenceLength - 1) >= len) {
+            // Not enough bytes to process yet - so store what we have and return
+            if (isFromServer) {
+#if defined(DEBUG_BIG5_PROCESSING)
+                qDebug().nospace() << "TBuffer::processBig5Sequence(...) Insufficent bytes in buffer to "
+                                      "complate Big5 sequence, need at least: "
+                                   << big5SequenceLength << " but we currently only have: " << bufferData.substr(pos).length() << " bytes (which we will store for next call to this method)...";
+#endif
+                mIncompleteSequenceBytes = bufferData.substr(pos);
+            }
+            return false; // Bail out
+        }
+
```

Comment:
You ought to validate the ranges of high and low byte pairs here and mark invalid pairs with the replacement character.

As it happens this is the reason I hadn't promulgated something like this in the past was because I could not see which "Big5" encoding A). Qt used (*since it is their decoder I am using - the reason we have to examine the byte stream is to deduce how many `QChar`s the incoming stream gets converted into so that we have the same number of `TChar`s*) and B). should be chosen from the those listed on the Wikipedia [Big5](https://en.wikipedia.org/wiki/Big5) page.

If you *can* do that it would be great - BTW to test the decoder you can use the `feedTriggers(`...`)` with the required ***byte*** values - provided that all the characters that the Big5 encodings use are within the Unicode BMP then you will never need to set `isNonBmpCharacter` to `true`...

Reply:
Can we do any validation - to detect any 2-byte values that the Qt decoder will reject? Ideally I guess it means picking through the tables in `./qtbase/src/corelib/codecs/qbig5codec.cpp` in the Qt source code and a bit of hackery with a spreadsheet or sommat to find the holes in the ranges of characters that are handled...